File: C++/310.cpp
class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
        vector<int> ans, indg(n);
        vector<vector<int>> g(n);
        for (auto e : edges){
            g[e[0]].push_back(e[1]);
            g[e[1]].push_back(e[0]);
            ++indg[e[0]];
            ++indg[e[1]];
        }
        queue<int> q;
        for (int i = 0; i < n; ++i) if (indg[i] == 1){
            q.push(i);
        }
        while(!q.empty()){
            ans.clear();
            for (int sz = size(q); sz > 0; --sz){
                int cur = q.front();
                q.pop();
                for (int next : g[cur]) if (--indg[next] == 1){
                    q.push(next);
                }
                ans.push_back(cur);
            }
        }
        return n==1?vector<int>{0}:ans;
    }
};

File: Java/310.java
class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        List<Integer> ans = new ArrayList<>();
        int[] ind = new int[n];
        Map<Integer, List<Integer>> map = new HashMap<>();
        for (int[] e : edges){
            map.computeIfAbsent(e[0], o->new ArrayList<>()).add(e[1]);
            map.computeIfAbsent(e[1], o->new ArrayList<>()).add(e[0]);
            ind[e[0]]++;
            ind[e[1]]++;
        }
        Queue<Integer> queue = new ArrayDeque<>();
        for (int i = 0; i < n; i++){
            if (ind[i]==1){
                queue.offer(i);
            }
        }
        while(!queue.isEmpty()){
            ans.clear();
            for (int sz = queue.size(); sz > 0; sz--){
                int cur = queue.poll();
                ans.add(cur);
                for (int next : map.get(cur)){
                    if (--ind[next]==1){
                        queue.offer(next);
                    }
                }
            }
        }
        return n==1?List.of(0):ans;
    }
}
